src/lib.rs

```diff
@@ -0,0 +1,371 @@
+//! Entry point into the task generation logic.
+//!
+//! This code will go through the entire evergreen configuration and create task definitions
+//! for any tasks that need to be generated. It will then add references to those generated
+//! tasks to any build variants to expect to run them.
+#![cfg_attr(feature = "strict", deny(missing_docs))]
+
+use std::{
+    collections::{HashMap, HashSet},
+    path::Path,
+    sync::{Arc, Mutex},
+};
+
+use anyhow::Result;
+use evg_config::{EvgConfigService, EvgConfigUtils, EvgConfigUtilsImpl, EvgProjectConfig};
+use lazy_static::lazy_static;
+use regex::Regex;
+use resmoke_proxy::ResmokeProxy;
+use shrub_rs::models::{
+    project::EvgProject,
+    task::{EvgTask, TaskRef},
+    variant::{BuildVariant, DisplayTask},
+};
+use task_name::remove_gen_suffix;
+use task_types::{
+    fuzzer_tasks::{FuzzerGenTaskParams, GenFuzzerService, GenFuzzerServiceImpl},
+    generated_suite::GeneratedSuite,
+    multiversion::MultiversionServiceImpl,
+};
+use tracing::{event, Level};
+
+mod evergreen_names;
+mod evg_config;
+mod resmoke_proxy;
+mod task_name;
+mod task_types;
+
+/// Directory to store the generated configuration in.
+const CONFIG_DIR: &str = "generated_resmoke_config";
+
+lazy_static! {
+    static ref EXPANSION_RE: Regex =
+        Regex::new(r"\$\{(?P<id>[a-zA-Z0-9_]+)(\|(?P<default>.*))?}").unwrap();
+}
+
+type GenTaskCollection = HashMap<String, Box<dyn GeneratedSuite>>;
+
+/// Collection of services needed to execution.
+#[derive(Clone)]
+pub struct Dependencies {
+    gen_task_service: Arc<dyn GenerateTasksService>,
+}
+
+impl Dependencies {
+    /// Create a new set of dependency instances.
+    ///
+    /// # Arguments
+    ///
+    /// * `evg_project_location` - Path to the evergreen project configuration yaml.
+    ///
+    /// # Returns
+    ///
+    /// A set of dependencies to run against.
+    pub fn new(evg_project_location: &Path) -> Result<Self> {
+        let discovery_service = Arc::new(ResmokeProxy::new());
```

Comment:
`Arc` has to do with memory management. It stands for "Atomic reference counter". It will let multiple threads access the data in the variable. We don't actually need it right now, but we will in a future ticket.